design-docs/012-python-namespaces

```diff
@@ -0,0 +1,216 @@
+- `storage`
+  - `storage.S3Options`
+  - `storage.Storage`
+  - `storage.StorageConcurrencySettings`
+  - `storage.StorageRetriesSettings`
+  - `storage.StorageSettings`
+  - `storage.AnyObjectStoreConfig`
+  - `storage.azure_storage`
+  - `storage.gcs_storage`
+  - `storage.gcs_store`
+  - `storage.http_store`
+  - `storage.in_memory_storage`
+  - `storage.local_filesystem_storage`
+  - `storage.local_filesystem_store`
+  - `storage.r2_storage`
+  - `storage.s3_storage`
+  - `storage.s3_store`
+  - `storage.tigris_storage`
+- `store`
+  - `store.IcechunkStore`
+- `xarray`
+  - `xarray.to_icechunk`
+
+New namespaces:
+
+- `config`
+  - `config.CachingConfig`
+  - `config.CompressionConfig`
+  - `config.CompressionAlgorithm`
+  - `config.ObjectStoreConfig`
+  - `config.RepositoryConfig`
+- `conflict`
+  - `conflict.BasicConflictSolver`
+  - `conflict.Conflict`
+  - `conflict.ConflictDetector`
+  - `conflict.ConflictSolver`
+  - `conflict.ConflictType`
+  - `conflict.VersionSelection`
+- `exceptions`
+  - `exceptions.ConflictError`
+  - `exceptions.IcechunkError`
+  - `exceptions.RebaseFailedError`
+- `logging`
+  - `logging.initialize_logs`,
+  - `logging.set_logs_filter`,
+- `manifests`
+  - `manifests.ManifestConfig`
+  - `manifests.ManifestFileInfo`
+  - `manifests.ManifestPreloadCondition`
+  - `manifests.ManifestPreloadConfig`
+  - `manifests.ManifestSplitCondition`
+  - `manifests.ManifestSplitDimCondition`
+  - `manifests.ManifestSplittingConfig`
+- `snapshots`
+  - `snapshots.SnapshotInfo`
+  - `snapshots.Diff`
+- `virtual`
+  - `virtual.VirtualChunkContainer`
+  - `virtual.VirtualChunkSpec`
+
+## Outstanding questions
+
+- It's generally a bit confusing that there are multiple ways to make most classes, e.g. `local_filesystem_store` and `Storage.new_local_filesystem`. Do we really need both?
+  - If we only kept `Storage` we could get rid of a _lot_ of functions in favour of methods...
+  - Alternatively should base classes that aren't really meant to be called by the user (e.g. `Storage`) be public API at all?
+- What's the difference between `storage.local_filesystem_storage` and `storage.local_filesystem_store`?
+  - Similarly how can we better disambiguate `storage.s3_storage` and `storage.s3_store`?
+- Is there a similar question for credentials?
+- Should cloud-provider-specific classes be further grouped together within `.storage`/`.credentials`?
+- Should all configs live together?
+  - Should `StorageSettings` really be a `StorageConfig`? Should it live with configs or in `storage`?
+  - Similarly should `S3Options` really be `S3Config`?
+- Should summary dataclasses such as `Diff` and `GCSummary` live in the top-level?
+- Is `distributed` meant to be a public module at all? Why is it's API defined in a different place? Should it be merged with `dask`?
+- Union types like `AnyCredential` are re-exported as top-level API for importing, but it seems like they are only useful for internal type-hinting purposes?
```

Comment:
Only for typing, but not sure if they are needed publicly or only privately

Reply:
If they are only needed for typing, and you don't expect users to make great use of them for typing, I think they should be private.